internal/jujucommands/jujucommands.go

```
@@ -0,0 +1,148 @@
+	// In runCmd, the writer (cmdWriter) is deferred to close only after runCmd completes.
+	// When sending the final error down cmdFinishedCh, if it's unbuffered,
+	// the send blocks because no goroutine is receiving yet.
+	// This blocks runCmd from completing, so the deferred writer.Close() never runs,
+	// meaning no EOF is sent to the reader, which keeps blocking forever.
+	// By buffering cmdFinishedCh, the send won't block, allowing runCmd to complete,
+	// the writer to close, and the reader to receive EOF and finish reading.
+	//
+	// There's no need to worry about it blocking, as we'll only ever send one value down
+	// once the command completes.
+	//
+	// We could alternatively manually call cmdWriter.Close() immediately after running the command.
+	cmdFinishedCh := make(chan error, 1)
+
+	go func() {
+		// Read cmdoutput.
+		lr := linereader.New(cmdReader)
+		for line := range lr.Ch {
+			outputCh <- outputLine{Line: line}
+		}
+
+		// Wait for cmd to finish.
+		err := <-cmdFinishedCh
+		if err != nil {
+			outputCh <- outputLine{Err: err}
+		}
+
+		close(outputCh)
+	}()
+	go runCmd(cmdCtx, cmdWriter, store, cmdFinishedCh, cmdAndArgs)
+
+	return outputCh, nil
+}
+
+// runCmd executes a Juju command using the provided context, client store, and command arguments.
+// It writes command output to the given io.PipeWriter and signals completion or error via cmdFinishedCh.
+// The function splits cmdAndArgs into arguments, constructs a Juju command, and runs it.
+// On non-zero exit code, it sends an error to cmdFinishedCh; otherwise, it signals successful completion.
+//
+// Note, the cmdAndArgs argument expects a fully-qualified command strings. I.e.,
+//
+//	"bootstrap lxd a --add-model=\"my-initial-model\""
+func runCmd(
+	cmdCtx *cmd.Context,
+	cmdWriter *io.PipeWriter,
+	store jujuclient.ClientStore,
+	cmdFinishedCh chan<- error,
+	cmdAndArgs string,
+) {
+	defer cmdWriter.Close()
+
+	jujuCmd := commands.NewJujuCommandWithStore(
```

Question on this, why don't we use `jujuCmd.Run(cmdCtx)` to execute the command? It seems that would return the error separately which might be easier to use.
It seems like we can also do `jujuCmd.Init(args)` `jujuCmd.SetFlags(flags)`.

---

We can, but there's two paths - A silent error and sometimes underlying details which are printed in a user friendly fashion. For example, it tries to bootstrap but cannot find the `PersonalCloudMetadata`, in this case the error would be from os.ReadFile which isn't helpful, it will have a nicely formatted written line just before though to be displayed to the user. So I don't think we need the underlying error for our use case honestly. We just care when the command has finished.

---

Gotcha. And then, should we at least use the `Init()` and `SetFlags` methods on the command so that we are not passing around a string that contains flags and arguments? Or is it preferred to pass around strings and then pass that string to cmd.Main()

---

I just checked to see if we could run the command manually without wrapping it in main and have our own registry, then using init & flags ourselves makes sense, but we can't because:
```go
func newBootstrapCommand() cmd.Command {
	command := &bootstrapCommand{}
	command.clock = jujuclock.WallClock
	command.CanClearCurrentModel = true
	return modelcmd.Wrap(command,
		modelcmd.WrapSkipModelFlags,
		modelcmd.WrapSkipDefaultModel,
	)
}
```
The command is private unfortunately. 

I did try setting flags manually using .Set, but because they aren't registered yet we'd have to register them otherwise we get an error of: ` no such option -format`

And... If we add the flag "format" and then run the command, we get a panic of: `option redefinition`.

So we're outta luck.